move zoneHVAC to subclass

brianlball opened the pull request.

jmarrec commented:
I apologize for being pedantic, but I would strongly advise when you aim to override a method you do it in two commits.

Commit 1: copy the original method, verbatim.
Commit 2: make the changes.

That way it's much easier to diff what was changed.

jmarrec committed:
Copy verbatim the original methods from Standards.ZoneHVACComponent.rb

zone_hvac_component_occupancy_ventilation_control and zone_hvac_component_apply_standard_controls

brianlball committed:
move zoneHVAC to subclass

jmarrec commented:
I replayed this PR by making two commits: copying the original method and the other is the actual changes.

jmarrec commented on lib/openstudio-standards/standards/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.ZoneHVACComponent.rb:
```diff
@@ -67,18 +67,24 @@ def zone_hvac_component_occupancy_ventilation_control(zone_hvac_component)
+    model = zone_hvac_component.model
+    building = model.getBuilding
+    acm_schedule_name = building.additionalProperties.getFeatureAsString('acm_fan_sch').get
+    occ_sch = model_add_schedule(model, acm_schedule_name)
```
Change 1 is here.

We own the fan schedule, which is in our standard data. so we should NOT rely on trying to get acm_fan_sch.

instead we should just map the building type to the right fan schedule and use that.

jmarrec replied:
Ok, so I added a new key on the spreadsheet standards tab **Space Types Thermostats**, column "**HVAC Operation Schedule**"
https://docs.google.com/spreadsheets/d/1SdiLBZlDr0wMbI7Kmfb-jXdnuoy-Io7-6DNmphshnnU/edit?disco=AAABPP1pqGo

I had said on https://github.com/NREL/openstudio-bem-to-surrogate-gem/pull/30#issuecomment-2123430331 that this would be needed.

cf b78aeb4d650bafb2325f6c37ae5a2318d1ee230a

jmarrec commented on lib/openstudio-standards/standards/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.ZoneHVACComponent.rb:
```diff
@@ -109,8 +115,8 @@ def zone_hvac_component_apply_standard_controls(zone_hvac_component)
 
-    # Standby mode occupancy control
-    return true unless zone_hvac_component.thermalZone.empty?
+    # Standby mode occupancy control - TODO check if it works right
+    return true if zone_hvac_component.thermalZone.empty?
```
Filed at:

* https://github.com/NREL/openstudio-standards/issues/1753

jmarrec committed:
new "hvac_operation_schedule" key

jmarrec committed:
Create a model_get_standards_data(model) method, make the space_type_get_standards_data use it + add test

jmarrec committed:
Missing includes for local testing

jmarrec committed:
Redo zone_hvac_component_occupancy_ventilation_control to lookup the schedule to use.

jmarrec committed:
Add `model_apply_acm_hvac_availability_schedule`. This is currently unused

I'm basically porting over the changes from https://github.com/NREL/openstudio-bem-to-surrogate-gem/pull/30

jmarrec commented on lib/openstudio-standards/standards/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.Model.rb:
```diff
@@ -166,4 +166,36 @@ def model_prm_baseline_system_number(_model, _climate_zone, area_type, _fuel_typ
 
     return sys_num
   end
+
+  # Returns standards data for selected model
+  # This will check the building primary type instead
+  #
+  # @param model [OpenStudio::Model::Model] the model
+  # @return [hash] hash of internal loads for different load types
+  def model_get_standards_data(model, throw_if_not_found: false)
```
New helper to get the standards data directly from a model.

jmarrec commented on lib/openstudio-standards/standards/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.SpaceType.rb:
```diff
@@ -26,28 +26,8 @@ def whole_building_space_type_name(model, primary_building_type)
-        raise msg
-      end
-      OpenStudio.logFree(OpenStudio::Warn, 'openstudio.standards.SpaceType', msg)
-      space_type_properties = {}
-    end
+    space_type_properties = model_get_standards_data(space_type.model, throw_if_not_found: throw_if_not_found)
```
space_type_get_standards_data will use that new helper, and extend_with_2007 if requested.

jmarrec commented on lib/openstudio-standards/standards/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/179d_ashrae_90_1_2007.ZoneHVACComponent.rb:
```diff
@@ -0,0 +1,139 @@
+    #                                                sch_name: "#{zone_hvac_component.name} Occ Sch",
+    #                                                occupied_percentage_threshold: occ_threshold)
+    ## NOTE: using 179d acm schedule
+    data = model_get_standards_data(zone_hvac_component.model, throw_if_not_found: true)
+    acm_fan_sch_name = data['hvac_operation_schedule']
+    zone_hvac_component.model.getBuilding.additionalProperties.setFeature('acm_fan_sch', acm_fan_sch_name)
+    acm_fan_sch = model_add_schedule(zone_hvac_component.model, acm_fan_sch_name)
+    zone_hvac_component.setSupplyAirFanOperatingModeSchedule(acm_fan_sch)
+    OpenStudio.logFree(OpenStudio::Info, 'openstudio.Standards.ZoneHVACComponent', "#{zone_hvac_component.name} has ventilation.  Setting fan operating mode schedule to ACM schedule '#{acm_fan_sch_name}'.")
```
Now I use the model_get_standards_data to grab the hvac_operation_schedule name from the hash.

jmarrec commented on lib/openstudio-standards/standards/ashrae_90_1/ashrae_90_1_2007/179d_ashrae_90_1_2007/data/179d_ashrae_90_1_2007.spc_typ.json:
```diff
@@ -52,6 +52,7 @@
       "infiltration_schedule_perimeter": null,
       "heating_setpoint_schedule": "Nonres_Heat_Sch",
       "cooling_setpoint_schedule": "Nonres_Cool_Sch",
+      "hvac_operation_schedule": "Nonres_HVAC_Sch",
```
New key!

jmarrec commented on test/90_1_general/test_179d.rb:
```diff
@@ -80,6 +80,18 @@ def test_model_get_primary_building_type
                  log_errors.first)
   end
 
+  def test_model_get_standards_data
+    assert_equal('Warehouse', standard.model_get_primary_building_type(model))
+
+    data = standard.model_get_standards_data(model)
+    assert_equal('179d-90.1-2007', data['template'])
+    assert_equal('Warehouse', data['building_type'])
+    assert_equal('WholeBuilding', data['space_type'])
+    assert_equal('ASHRAE 90.1-2007', data['lighting_standard'])
+    assert_equal('WholeBuilding', data['lighting_primary_space_type'])
+    assert_equal('Warehouse', data['lighting_secondary_space_type'])
+    assert_equal('Nonres_HVAC_Sch', data['hvac_operation_schedule'])
+  end
```
test for new helper

jmarrec commented on test/90_1_general/test_179d.rb:
```diff
@@ -257,4 +269,28 @@ def test_space_type_get_standards_data
     end
   end
 
+  def test_zone_hvac_component_apply_standard_controls
+    fan = OpenStudio::Model::FanSystemModel.new(model)
+    heating_coil = OpenStudio::Model::CoilHeatingWater.new(model)
+    cooling_coil = OpenStudio::Model::CoilCoolingWater.new(model)
+    four_pipe_fan_coil = OpenStudio::Model::ZoneHVACFourPipeFanCoil.new(model, model.alwaysOnDiscreteSchedule, fan, cooling_coil, heating_coil)
+    four_pipe_fan_coil.addToThermalZone(model.getThermalZones.first)
+
+    assert_empty(four_pipe_fan_coil.supplyAirFanOperatingModeSchedule)
+
+    four_pipe_fan_coil.setMaximumOutdoorAirFlowRate(0.0)
+    standard.zone_hvac_component_apply_standard_controls(four_pipe_fan_coil.to_ZoneHVACComponent.get)
+    assert_empty(four_pipe_fan_coil.supplyAirFanOperatingModeSchedule)
+
+    four_pipe_fan_coil.autosizeMaximumOutdoorAirFlowRate
+    assert_empty(four_pipe_fan_coil.maximumOutdoorAirFlowRate)
+    assert four_pipe_fan_coil.isMaximumOutdoorAirFlowRateAutosized
+    standard.zone_hvac_component_apply_standard_controls(four_pipe_fan_coil.to_ZoneHVACComponent.get)
+    refute_empty(four_pipe_fan_coil.supplyAirFanOperatingModeSchedule)
+
+    four_pipe_fan_coil.resetSupplyAirFanOperatingModeSchedule
+    four_pipe_fan_coil.setMaximumOutdoorAirFlowRate(0.1)
+    standard.zone_hvac_component_apply_standard_controls(four_pipe_fan_coil.to_ZoneHVACComponent.get)
+    refute_empty(four_pipe_fan_coil.supplyAirFanOperatingModeSchedule)
+  end
```
Test for the ZoneHVAC override.

jmarrec reviewed (commented):
@brianlball redid your PR, please check.

jmarrec commented:
@brianlball I have added extra changes to a PR that points to this branch in https://github.com/jmarrec/openstudio-standards/pull/12

brianlball commented:
@JanghyunJK @anchapin

jmarrec committed:
Merge pull request #12 from jmarrec/model_apply_acm_hvac_availability_schedule

Add `model_apply_acm_hvac_availability_schedule`.

jmarrec commented:
Still wanting a review, but I'm merging this for now. I've written tests and I'm confident I'm respecting the intent.
